Client/Helpers/FxALoginHelper.swift

```diff
@@ -0,0 +1,210 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Account
+import Deferred
+import Foundation
+import Shared
+import SwiftyJSON
+import XCGLogger
+
+private let applicationDidRequestUserNotificationPermissionPrefKey = "applicationDidRequestUserNotificationPermissionPrefKey"
+
+private let log = Logger.browserLogger
+
+protocol FxAPushLoginDelegate : class {
+    func accountLoginDidFail()
+
+    func accountLoginDidSucceed(withFlags flags: FxALoginFlags)
+}
+
+/// Small struct to keep together the immediately actionable flags that the UI is likely to immediately
+/// following a successful login. This is not supposed to be a long lived object.
+struct FxALoginFlags {
+    let pushEnabled: Bool
+    let verified: Bool
+}
+
+/// This class manages the from successful login for FxAccounts to 
+/// asking the user for notification permissions, registering for 
+/// remote push notifications (APNS), registering for WebPush notifcations
+/// then creating an account and storing it in the profile.
+class FxALoginHelper {
+    static var sharedInstance: FxALoginHelper = {
+        return FxALoginHelper()
+    }()
+
+    weak var delegate: FxAPushLoginDelegate?
+
+    fileprivate weak var profile: Profile?
+
+    fileprivate var account: FirefoxAccount!
+
+    fileprivate var accountVerified: Bool!
+
+    // This should be called when the application has started.
+    // This configures the helper for logging into Firefox Accounts, and
+    // if already logged in, checking if anything needs to be done in response
+    // to changing of user settings and push notifications.
+    func application(_ application: UIApplication, didLoadProfile profile: Profile) {
+        self.profile = profile
+
+        guard let account = profile.getAccount() else {
+            // There's no account, no further action.
+            return loginDidFail()
+        }
+
+        guard AppConstants.MOZ_FXA_PUSH else {
+            return loginDidSucceed()
+        }
+
+        if let _ = account.pushRegistration {
+            // We have an account, and it's already registered for push notifications.
+            return loginDidSucceed()
+        }
+
+        // Now: we have an account that does not have push notifications set up.
+        // however, we need to deal with cases of asking for permissions too frequently.
+        let asked = profile.prefs.boolForKey(applicationDidRequestUserNotificationPermissionPrefKey) ?? true
+        let permitted = application.currentUserNotificationSettings!.types != .none
+
+        // If we've never asked(*), then we should probably ask.
+        // If we've asked already, then we should not ask again.
+        // TODO: add UI to tell the user to go flip the Setting app.
```

Comment:
Thought: When we do add in the UI for asking the user to accept pushes, do we want the event to show the UI to be driven by the `didLoadProfile` event or initiated by part of the UI? For example, if want to show the UI at a particular place in the app, it might be easier to have the view controller request to show the UI and ask the helper if we need to vs having the helper auto-show when the profile loaded.